packages/cozy-client/src/CozyClient.js

```diff
@@ -821,21 +831,22 @@ client.query(Q('io.cozy.bills'))`)
       }
       const shouldFetch = options.fetchPolicy(existingQuery)
       if (!shouldFetch) {
-        return
-      }
-    }
-    // If we already have the same query in loading state, no
-    // need to refetch it
-    if (existingQuery && Object.keys(existingQuery).length > 0) {
-      if (existingQuery.fetchStatus === 'loading') {
-        return
+        return this.getQueryFromState(queryId, {
+          hydrated: get(options, 'hydrated', true),
+          singleDocData: get(options, 'singleDocData', false)
+        })
       }
     }
     this.ensureQueryExists(queryId, queryDefinition, options)
 
     try {
       this.dispatch(loadQuery(queryId))
-      const response = await this.requestQuery(queryDefinition)
+
+      const response = await this._promiseCache.exec(
+        () => this.requestQuery(queryDefinition),
+        () => stringify(queryDefinition)
+      )
+
```

Comment:
Below, the returned result is a response. I think it should also be made directly from the state and not be a Response object, this way, both code path return the same thing. I fear that the response and the hydrated query result are not exactly the same thing. Unfortunately, the QueryResult type here is too simple and is not protecting us.

Reply:
I had the same fear. Especially for relationships. After few tests, I had the same things, but you're right having only one path to return something is the way to go. 

I've reverted this modification, will create an issue and work on it later since it appears to break a lot of tests at first.